tests/Cache/FileHandlerTest.php

```diff
@@ -145,6 +148,76 @@ public function testWriteWritesCache()
         $this->assertSame($cache->toJson(), $actualCacheJson);
     }
 
+    public function testWriteCacheToDirectory()
+    {
+        $dir = __DIR__.'/../Fixtures/cache-file-handler';
+
+        $handler = new FileHandler($dir);
+
+        $this->setExpectedExceptionRegExp(
+            'Symfony\Component\Filesystem\Exception\IOException',
+            sprintf('#^%s$#', preg_quote('Cannot write cache file "'.realpath($dir).'" as the location exists as directory.', '#')
+        ));
+
+        $handler->write(new Cache(new Signature(
+            PHP_VERSION,
+            '2.0',
+            array(
+                'foo',
+                'bar',
+            )
+        )));
+    }
+
+    public function testWriteCacheToNonWriteableFile()
+    {
+        $file = __DIR__.'/../Fixtures/cache-file-handler/cache-file';
+        if (is_writable($file)) {
+            $this->markTestSkipped(sprintf('File "%s" must be not writeable for this tests.', realpath($file)));
+
+            return;
+        }
+
+        $handler = new FileHandler($file);
+
+        $this->setExpectedExceptionRegExp(
+            'Symfony\Component\Filesystem\Exception\IOException',
+            sprintf('#^%s$#', preg_quote('Cannot write to file "'.realpath($file).'" as it is not writable.', '#')
+        ));
+
+        $handler->write(new Cache(new Signature(
+            PHP_VERSION,
+            '2.0',
+            array(
+                'foo',
+                'bar',
+            )
+        )));
+    }
+
+    public function testWriteCacheFilePermissions()
+    {
+        $file = __DIR__.'/../Fixtures/cache-file-handler/rw_cache.test';
+        @unlink($file);
+
+        $this->assertFileNotExists($file);
+
+        $handler = new FileHandler($file);
+        $handler->write(new Cache(new Signature(
+            PHP_VERSION,
+            '2.0',
+            array(
+                'foo',
+                'bar',
+            )
+        )));
+
+        $this->assertFileExists($file);
+        $this->assertFilePermissionMask(0666, $file);
+
+        @unlink($file);
```

Comment:
This is not enough, that's probably why you had to try drop this file also in top of the method.
File shall be registered to be removed even if php will crash and not execute this line at all, use `FileRemoval` to do it

Reply:
`FileRemoval` works based on `register_shutdown_function`, therefor `removed even if php will crash` is not true. When PHP exists on a fatal like out-op-memory the `register_shutdown_function` are not quarantined to be executed.

Reply:
`@unlink` will also not handle `out-op-memory` cases, yet, `FileRemoval` will handle more cases than just `@unlink` ?

Reply:
```
@unlink will also not handle out-op-memory cases, yet, FileRemoval will handle more cases than just @unlink ?
```

Don't think so, my setup always throws away the file if it exists before testing and tries to throw it away at the end of it.

Reply:
change
```diff
+ throw new \Exception();
  @unlink($file);
```

is file removed? no ;)
with `FileRemoval`, it would. that's my point

Reply:
Using `FileRemoval` might help for an exception, it doesn't for uncontrolled stopping of the process. My point is that my version doesn't care about any of that.

Reply:
but then you have weird unlinking in the beginning of test ;)

Reply:
From your POV ;)
I like to _setup_ my test at the _start_,
so in this test I don't want to have that file so I throw it away.